File: packages/proof/src/canonical.ts
/** Restricted canonical encoding for pilot records; no policy or authorization decision. */
import { createHash } from 'node:crypto';
export function canonicalBytes(value: unknown): Buffer {
  let budget = 65536;
  const charge = (n: number) => { budget -= n; if (budget < 0) throw new Error('Canonical record exceeds validation budget'); };
  const encode = (item: unknown, depth: number): string => {
    charge(8);
    if (depth > 8) throw new Error('Canonical record exceeds nesting limit');
    if (item === null || typeof item === 'boolean') return JSON.stringify(item);
    if (typeof item === 'number' && Number.isSafeInteger(item) && !Object.is(item, -0)) return String(item);
    if (typeof item === 'string' && item.length <= 4096 && /^[ -~]*$/.test(item) && !/[\r\n]/.test(item)) {
      charge(item.length); return JSON.stringify(item);
    }
    if (Array.isArray(item) && item.length <= 256) {
      // Reject sparse arrays; JSON.stringify would silently turn holes into null.
      const parts: string[] = [];
      for (let i = 0; i < item.length; i++) {
        if (!Object.hasOwn(item, i)) throw new Error('Sparse canonical array');
        parts.push(encode(item[i], depth + 1));
      }
      return '[' + parts.join(',') + ']';
    }
    if (item && typeof item === 'object' && Object.getPrototypeOf(item) === Object.prototype) {
      if (Object.values(Object.getOwnPropertyDescriptors(item)).some(d => d.get || d.set)) throw new Error('Canonical accessors are forbidden');
      const entries = Object.entries(item);
      if (entries.length > 64 || Reflect.ownKeys(item).length !== entries.length) throw new Error('Invalid canonical object');
      return '{' + entries.sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0).map(([key, child]) => {
        if (!/^[ -~]{1,128}$/.test(key) || /[\r\n]/.test(key)) throw new Error('Invalid canonical record key');
        charge(key.length + 3);
        return JSON.stringify(key) + ':' + encode(child, depth + 1);
      }).join(',') + '}';
    }
    throw new Error('Unsupported canonical record value');
  };
  const bytes = Buffer.from(encode(value, 0), 'ascii');
  if (bytes.length > 65536) throw new Error('Canonical record exceeds 64 KiB');
  return bytes;
}
export function recordDigest(domain: string, value: unknown): string {
  if (!/^clearproof\/[a-z-]+\/v[1-9][0-9]*$/.test(domain) || /[\r\n]/.test(domain)) throw new Error('Invalid commitment domain');
  return createHash('sha256').update(domain, 'ascii').update(Buffer.from([0])).update(canonicalBytes(value)).digest('hex');
}

File: src/protocol/canonical.py
"""Bounded canonical JSON for versioned pilot records, not arbitrary Unicode JSON."""

from __future__ import annotations

import hashlib
import json
import re
from typing import Any

MAX_RECORD_BYTES = 65536


def canonical_bytes(value: Any) -> bytes:
    """ASCII strings, safe integers, booleans, null, arrays and objects only.

    This restricted profile avoids floating point and Unicode sorting ambiguity.
    It is intentionally not a general implementation of RFC 8785.
    """

    budget = MAX_RECORD_BYTES

    def charge(amount: int) -> None:
        nonlocal budget
        budget -= amount
        if budget < 0:
            raise ValueError("Canonical record exceeds validation budget")

    def validate(item: Any, depth: int) -> None:
        charge(8)
        if depth > 8:
            raise ValueError("Canonical record exceeds nesting limit")
        if item is None or type(item) is bool:
            return
        if type(item) is int and abs(item) <= 2**53 - 1:
            return
        if type(item) is str and len(item) <= 4096 and all(32 <= ord(c) <= 126 for c in item):
            charge(len(item))
            return
        if type(item) is list and len(item) <= 256:
            for child in item:
                validate(child, depth + 1)
            return
        if type(item) is dict and len(item) <= 64:
            for key, child in item.items():
                if type(key) is not str or not re.fullmatch(r"[ -~]{1,128}", key):
                    raise ValueError("Invalid canonical record key")
                charge(len(key) + 3)
                validate(child, depth + 1)
            return
        raise ValueError("Unsupported canonical record value")

    validate(value, 0)
    result = json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=True, allow_nan=False).encode(
        "ascii"
    )
    if len(result) > MAX_RECORD_BYTES:
        raise ValueError("Canonical record exceeds 64 KiB")
    return result


def record_digest(domain: str, value: Any) -> str:
    if not re.fullmatch(r"clearproof/[a-z-]+/v[1-9][0-9]*", domain):
        raise ValueError("Invalid commitment domain")
    return hashlib.sha256(domain.encode("ascii") + b"\x00" + canonical_bytes(value)).hexdigest()
